lib/xapi-fdcaps/operations.ml

```diff
@@ -0,0 +1,206 @@
+  let ro, wo = Unix.pipe ~cloexec:true () in
+  (make_ro_exn kind ro, make_wo_exn kind wo)
+
+let socketpair domain typ proto =
+  let kind = `sock in
+  let fd1, fd2 = Unix.socketpair ~cloexec:true domain typ proto in
+  (make_rw_exn kind fd1, make_rw_exn kind fd2)
+
+let openfile_ro kind path flags =
+  make_ro_exn kind
+  @@ Unix.openfile path (Unix.O_RDONLY :: Unix.O_CLOEXEC :: flags) 0
+
+let openfile_rw ?custom_ftruncate kind path flags =
+  make_rw_exn ?custom_ftruncate kind
+  @@ Unix.openfile path (Unix.O_RDWR :: Unix.O_CLOEXEC :: flags) 0
+
+let openfile_wo kind path flags =
+  make_wo_exn kind
+  @@ Unix.openfile path (Unix.O_WRONLY :: Unix.O_CLOEXEC :: flags) 0
+
+let creat path flags perm =
+  make_rw_exn `reg
+  @@ Unix.openfile path
+       (Unix.O_RDWR :: Unix.O_CREAT :: Unix.O_EXCL :: Unix.O_CLOEXEC :: flags)
+       perm
+
+let dev_null_out () = openfile_wo `chr "/dev/null" []
+
+let dev_null_in () = openfile_ro `chr "/dev/null" []
+
+let dev_zero () = openfile_ro `chr "/dev/zero" []
+
+let shutdown_recv t =
+  Unix.shutdown (Safefd.unsafe_to_file_descr_exn t.fd) Unix.SHUTDOWN_RECEIVE
+
+let shutdown_send t =
+  Unix.shutdown (Safefd.unsafe_to_file_descr_exn t.fd) Unix.SHUTDOWN_SEND
+
+let shutdown_all t =
+  Unix.shutdown (Safefd.unsafe_to_file_descr_exn t.fd) Unix.SHUTDOWN_ALL
+
+let ftruncate t size =
+  match t.custom_ftruncate with
+  | None ->
+      Unix.LargeFile.ftruncate (Safefd.unsafe_to_file_descr_exn t.fd) size
+  | Some f ->
+      f size
+
+let lseek t off whence =
+  Unix.LargeFile.lseek (Safefd.unsafe_to_file_descr_exn t.fd) off whence
+
+let read t buf off len =
+  Unix.read (Safefd.unsafe_to_file_descr_exn t.fd) buf off len
+
+let single_write_substring t buf off len =
+  Unix.single_write_substring (Safefd.unsafe_to_file_descr_exn t.fd) buf off len
+
+let set_nonblock t = Unix.set_nonblock (Safefd.unsafe_to_file_descr_exn t.fd)
+
+let clear_nonblock t = Unix.clear_nonblock (Safefd.unsafe_to_file_descr_exn t.fd)
+
+let with_tempfile ?size () f =
+  let name, ch =
+    Filename.open_temp_file ~mode:[Open_binary] "xapi_fdcaps" "tmp"
+  in
+  let finally () =
+    close_out_noerr ch ;
+    try Unix.unlink name with Unix.Unix_error (_, _, _) -> ()
+  in
+  let@ () = Fun.protect ~finally in
+  let t = ch |> Unix.descr_of_out_channel |> make_wo_exn `reg in
+  let@ t = with_fd t in
+  size |> Option.iter (fun size -> ftruncate t size) ;
+  f (name, t)
+
+let check_output cmd args =
+  let cmd = Filename.quote_command cmd args in
+  let ch = Unix.open_process_in cmd in
+  let finally () =
+    try
+      let (_ : Unix.process_status) = Unix.close_process_in ch in
+      ()
+    with _ -> ()
+  in
+  Fun.protect ~finally @@ fun () ->
+  let out = In_channel.input_all ch |> String.trim in
+  match Unix.close_process_in ch with
+  | Unix.WEXITED 0 ->
+      out
+  | _ ->
+      failwith (Printf.sprintf "%s exited nonzero" cmd)
+
+let with_temp_blk ?(sector_size = 512) ?delay_read_ms:_ ?delay_write_ms:_ name f
```

Comment:
These delays should be respected, or removed from the interface

Reply:
The delays can be implemented with [dmsetup delay ](https://docs.kernel.org/admin-guide/device-mapper/delay.html). I can remove `with_temp_blk` from here and add it back once that is implemented